### PR TITLE
Refactor: move network related errors to MetaNetworkError, and raft related errors to MetaRaftError

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Move network related errors to MetaNetworkError, and raft related errors to MetaRaftError,which is sub types of MetaError.

## Changelog

- Improvement

## Related Issues

Fixes https://github.com/datafuselabs/databend/issues/4083

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/meta/types/src/meta_errors_into.rs [hunk 1]
```diff
@@ -39,6 +41,18 @@ impl From<MetaError> for ErrorCode {
...
+        MetaError::MetaNetworkError(e)
+    }
+}
+
+impl From<MetaRaftError> for MetaError {
+    fn from(e: MetaRaftError) -> Self {
+        MetaError::MetaRaftError(e)
+    }
+}
+
```
Comment (drmingdrmer) [inline common/meta/types/src/meta_errors_into.rs:55]:
```
MetaError {
   MetaRaftError(#[from] MetaRaftError)
}
```
Will automatically build a impl of `From<>`

File: common/meta/types/src/meta_errors_into.rs [hunk 2]
```diff
@@ -39,24 +38,12 @@ impl From<MetaError> for ErrorCode {
...
-    fn from(e: MetaStorageError) -> Self {
-        MetaError::MetaStorageError(e)
-    }
-}
-
 impl From<serde_json::Error> for MetaError {
     fn from(error: serde_json::Error) -> Self {
         MetaError::MetaStorageError(error.into())
     }
 }
```
Comment (drmingdrmer) [inline common/meta/types/src/meta_errors_into.rs:46]: If MetaStorageError is a superset of serde json Error, maybe MetaError itself does not need to include serde json Error.

If it is true, maybe we could fix it in next PR 🤔

File: common/meta/types/src/meta_network_errors.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#[derive(Error, Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub enum MetaNetworkError {
+    #[error(transparent)]
+    ConnectionError(#[from] ConnectionError),
+
+    #[error("{0}")]
+    GetNodeAddrError(String),
+
+    #[error("{0}")]
+    UnknownNode(String),
```
Comment (drmingdrmer) [inline common/meta/types/src/meta_network_errors.rs:30]: Do these two errors caused by the same reason that a node is not found in the local store?

If it is true, we could merge these two into one.

File: common/meta/types/src/meta_raft_errors.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    ForwardToLeader(#[from] ForwardToLeader),
+
+    #[error(transparent)]
+    ChangeMembershipError(#[from] ChangeMembershipError),
+
+    #[error("{0}")]
+    ConsistentReadError(String),
+
+    #[error("{0}")]
+    ClientWriteError(String),
```
Comment (drmingdrmer) [inline common/meta/types/src/meta_raft_errors.rs:36]: It seems there is only one error will be wrapped in `ClientWriteError`: `openraft::Fatal`.
What about naming it with `RaftFatal` or just `Fatal`?

BTW to wrap an external error, using `AnyError` with provide more info for debugging than using a `String`.

File: metasrv/src/meta_service/raftmeta.rs [hunk 5]
```diff
@@ -201,7 +203,14 @@ impl MetaNode {
...
 
         let addr_str = addr.to_string();
-        let addr = addr.parse::<std::net::SocketAddr>()?;
+        let ret = addr.parse::<std::net::SocketAddr>();
+        let addr = match ret {
+            Ok(addr) => addr,
+            Err(e) => {
+                let err: MetaNetworkError = e.into();
+                return Err(err.into());
+            }
```
Comment (drmingdrmer) [inline metasrv/src/meta_service/raftmeta.rs:212]: If there is a double `into()`, maybe the method return value type is too general.

IMHO, a `MetaNetworkError` may be enough for this `start_grpc()` method.

**Review by drmingdrmer — COMMENTED** [review]
It's became more clear! :DDD Love it.

**Review by drmingdrmer — APPROVED** [review]
LGTM:DDD

**Review by ariesdevil — APPROVED** [review]

**Pull request merged**